File: tools/tf-types-ts/src/core/instance-id.ts
import type { ActorType } from "../generated/_common.js";
import { ACTOR_TYPES, ActorIdParseError, formatActorId, splitScheme } from "./actor-id.js";

export class InstanceIdParseError extends ActorIdParseError {}

const ACTOR_TYPE_SET: ReadonlySet<string> = new Set<string>(ACTOR_TYPES);

export interface ParsedInstanceId {
  readonly type: ActorType;
  readonly actorPath: string;
  readonly instancePath: string;
  readonly raw: string;
}

/**
 * Parse an instance URI of the form `tf:instance:<type>:<actor-path>/<instance-path>`.
 *
 * The split between actor-path and instance-path is the last `/` in the URI.
 * Both halves must be non-empty.
 */
export function parseInstanceId(s: string): ParsedInstanceId {
  if (typeof s !== "string") throw new InstanceIdParseError(`expected string, got ${typeof s}`);
  const parts = splitScheme(s);
  if (!parts || parts.kind !== "instance") {
    throw new InstanceIdParseError(`expected tf:instance:<type>:<actor-path>/<instance-path>, got ${JSON.stringify(s)}`);
  }
  if (!ACTOR_TYPE_SET.has(parts.typeSegment)) {
    throw new InstanceIdParseError(`unknown actor type: ${parts.typeSegment}`);
  }
  const split = parts.path.lastIndexOf("/");
  if (split <= 0) {
    throw new InstanceIdParseError(`instance id must contain '/' separating actor and instance: ${s}`);
  }
  const actorPath = parts.path.slice(0, split);
  const instancePath = parts.path.slice(split + 1);
  if (!actorPath || !instancePath) {
    throw new InstanceIdParseError(`empty actor or instance segment in ${s}`);
  }
  return { type: parts.typeSegment as ActorType, actorPath, instancePath, raw: s };
}

export function formatInstanceId(p: { type: ActorType; actorPath: string; instancePath: string }): string {
  if (!ACTOR_TYPE_SET.has(p.type)) throw new InstanceIdParseError(`unknown actor type: ${p.type}`);
  if (!p.actorPath) throw new InstanceIdParseError("instance id actor-path is empty");
  if (!p.instancePath) throw new InstanceIdParseError("instance id instance-path is empty");
  return `tf:instance:${p.type}:${p.actorPath}/${p.instancePath}`;
}

export function toActorId(instanceId: string): string {
  const parsed = parseInstanceId(instanceId);
  return formatActorId({ type: parsed.type, path: parsed.actorPath });
}

File: crates/tf-types/src/instance_id.rs
//! Instance-URI parser mirroring `tools/tf-types-ts/src/core/instance-id.ts`.

use crate::actor_id::{
    actor_type_to_str, format_actor_id, parse_actor_type, split_scheme, ActorIdParseError,
};
use crate::generated::common::ActorType;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ParsedInstanceId {
    pub actor_type: ActorType,
    pub actor_path: String,
    pub instance_path: String,
    pub raw: String,
}

pub fn parse_instance_id(s: &str) -> Result<ParsedInstanceId, ActorIdParseError> {
    let parts = split_scheme(s).ok_or_else(|| ActorIdParseError::MalformedScheme(s.to_owned()))?;
    if parts.kind != "instance" {
        return Err(ActorIdParseError::WrongKind(parts.kind.to_owned()));
    }
    let actor_type = parse_actor_type(parts.type_segment)
        .ok_or_else(|| ActorIdParseError::UnknownType(parts.type_segment.to_owned()))?;
    let split = parts.path.rfind('/').ok_or(ActorIdParseError::EmptyPath)?;
    if split == 0 {
        return Err(ActorIdParseError::EmptyPath);
    }
    let actor_path = &parts.path[..split];
    let instance_path = &parts.path[split + 1..];
    if actor_path.is_empty() || instance_path.is_empty() {
        return Err(ActorIdParseError::EmptyPath);
    }
    Ok(ParsedInstanceId {
        actor_type,
        actor_path: actor_path.to_owned(),
        instance_path: instance_path.to_owned(),
        raw: s.to_owned(),
    })
}

pub fn format_instance_id(
    actor_type: &ActorType,
    actor_path: &str,
    instance_path: &str,
) -> Result<String, ActorIdParseError> {
    if actor_path.is_empty() || instance_path.is_empty() {
        return Err(ActorIdParseError::EmptyPath);
    }
    Ok(format!(
        "tf:instance:{}:{}/{}",
        actor_type_to_str(actor_type),
        actor_path,
        instance_path
    ))
}

pub fn to_actor_id(instance_id: &str) -> Result<String, ActorIdParseError> {
    let parsed = parse_instance_id(instance_id)?;
    format_actor_id(&parsed.actor_type, &parsed.actor_path)
}
